api/src/main/java/org/openmrs/api/db/hibernate/HibernateMedicationDispenseDAO.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.db.hibernate;
+
+import org.hibernate.SessionFactory;
+import org.openmrs.MedicationDispense;
+import org.openmrs.api.db.MedicationDispenseDAO;
+import org.openmrs.parameter.MedicationDispenseCriteria;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Hibernate implementation of the MedicationDispenseDAO
+ * @since 2.6.0
+ * @see MedicationDispenseDAO
+ */
+public class HibernateMedicationDispenseDAO implements MedicationDispenseDAO {
+	
+	private SessionFactory sessionFactory;
+	
+	public void setSessionFactory(SessionFactory sessionFactory) {
+		this.sessionFactory = sessionFactory;
+	}
+
+	@Override
+	public MedicationDispense getMedicationDispense(Integer medicationDispenseId) {
+		return sessionFactory.getCurrentSession().get(MedicationDispense.class, medicationDispenseId);
+	}
+
+	@Override
+	public MedicationDispense getMedicationDispenseByUuid(String uuid) {
+		return sessionFactory.getCurrentSession()
+			.createQuery("select md from MedicationDispense md where md.uuid = :uuid", MedicationDispense.class)
+			.setParameter("uuid", uuid)
+			.uniqueResult();
+	}
+
+	@Override
+	public List<MedicationDispense> getMedicationDispenseByCriteria(MedicationDispenseCriteria criteria) {
+		CriteriaBuilder criteriaBuilder = sessionFactory.getCurrentSession().getCriteriaBuilder();
+		CriteriaQuery<MedicationDispense> criteriaQuery = criteriaBuilder.createQuery(MedicationDispense.class);
+		Root<MedicationDispense> md = criteriaQuery.from(MedicationDispense.class);
+		List<Predicate> predicates = new ArrayList<>();
+		if (criteria.getPatient() != null) {
+			predicates.add(criteriaBuilder.equal(md.get("patient"), criteria.getPatient()));
+		}
+		if (criteria.getEncounter() != null) {
+			predicates.add(criteriaBuilder.equal(md.get("encounter"), criteria.getEncounter()));
+		}
+		if (criteria.getDrugOrder() != null) {
+			predicates.add(criteriaBuilder.equal(md.get("drugOrder"), criteria.getDrugOrder()));
+		}
+		if (!criteria.isIncludeVoided()) {
+			predicates.add(criteriaBuilder.equal(md.get("voided"), false));
+		}
+		if (predicates.size() > 0) {
+			criteriaQuery.where(predicates.toArray(new Predicate[]{}));
+		}
+		// TODO: What should we order by here for consistency?
```

Comment:
I feel like we should have a consistent and sensible ordering of results - we do in most of the service methods.  Usually there is a straightforward date field that can be used for this, but I'm not sure what that is in this case.  We should look at the date fields in general and clarify what one can expect there - there are encounter dates (optional), order dates (optional), and a few optional dates on the medicationdispense for when certain things happen.  How do we plan to think about Medication Dispense items sorted by date?  @mogoodrich / @dkayiwa  / @ibacher / @bmamlin

Reply:
Good question.  No immediate strong opinions.  Date prepared seems like my gut answer, but, as you pointed out, it's optional.  I'd be okay with making it or another date mandatory if we want to, it order to give consistent sorting... interested in what others think.

Reply:
I'm going to address this by ordering by "medicationDispenseId" as a starting point, which will at least ensure consistent results each time this method is called, and is a reasonable enough choice.  We can modify as we feel necessary moving forward.

Reply:
Hmmm... that's a good question. There's a workflow consideration here, I think. To me, it would make the most sense to rely on the date handed over since it's a "dispense" event and that would be the field that stores "when the medication was given to the patient", which is probably the most clinically-relevant field. However, I can see a case where a pharmacist might want to setup all the data for a medication dispense before the drug was actually handed over... if that's a use-case we want to support than date prepared seems to make the most sense.

There's no reason we can't add constraints on top of the existing FHIR constraints, e.g., to make one or the other date fields required. FHIR resources are intended to be a bit general so they can be widely adopted with the idea that different business rules will actually restrict the contents of the resource.